python/lsst/obs/base/gen2to3/convertRepo.py

```diff
@@ -475,66 +484,68 @@ def run(self, root: str, *,
-            # Allow registration to fail on the assumption that this means
-            # we are reusing a butler
-            try:
-                self.instrument.register(self.registry)
-            except Exception:
-                pass
-
-        # Make and prep converters for all Gen2 repos.  This should not modify
-        # the Registry database or filesystem at all, though it may query it.
-        # The prep() calls here will be some of the slowest ones, because
-        # that's when we walk the filesystem.
+        # Make converters for all Gen2 repos.
         converters = []
         rootConverter = RootRepoConverter(task=self, root=root, subset=subset)
-        rootConverter.prep()
         converters.append(rootConverter)
-
         for calibRoot, run in calibs.items():
             if not os.path.isabs(calibRoot):
                 calibRoot = os.path.join(rootConverter.root, calibRoot)
             converter = CalibRepoConverter(task=self, root=calibRoot, run=run,
                                            mapper=rootConverter.mapper,
                                            subset=rootConverter.subset)
-            converter.prep()
             converters.append(converter)
-
         for spec in reruns:
             runRoot = spec.path
             if not os.path.isabs(runRoot):
                 runRoot = os.path.join(rootConverter.root, runRoot)
             converter = StandardRepoConverter(task=self, root=runRoot, run=spec.runName,
                                               subset=rootConverter.subset)
-            converter.prep()
             converters.append(converter)
 
-        # Actual database writes start here.  We can't wrap these sanely in
-        # transactions (yet) because we keep initializing new Butler instances
-        # just so we can write into new runs/collections, and transactions
-        # are managed at the Butler level (DM-21246 should let us fix this).
+        # Register the instrument if we're configured to do so.
+        if self.config.doRegisterInstrument:
+            # Allow registration to fail on the assumption that this means
+            # we are reusing a butler
+            try:
+                self.instrument.register(self.registry)
+            except Exception:
+                pass
+
+        # Run raw ingest (does nothing if we weren't configured to convert the
+        # 'raw' dataset type).
+        rootConverter.runRawIngest()
+
+        # Write curated calibrations to all calibration repositories.
+        # Add new collections to the list of collections the butler was
+        # initialized to pass to DefineVisitsTask, to deal with the (likely)
+        # case the only 'camera' dataset in the repo will be one we're adding
+        # here.
+        if self.config.doWriteCuratedCalibrations:
+            for _, run in calibs.items():
+                butler3 = Butler3(butler=self.butler3, run=run)
+                self.instrument.writeCuratedCalibrations(butler3)
```

Comment:
Good point about this being wasteful; I hadn't considered that.  But it's also hard to fix, because `writeCuratedCalibration` (or `Instrument` more generally) doesn't provide a way to ask which datasets it actually adds.

I've also started thinking that `ConvertRepoTask` would be a lot cleaner if we could remove the instrument registration and curated-calibration responsibilities from it once we have command-line utilities for those, and maybe the raw ingest ones, too (I'm less sure about the latter because it contains some logic for selecting which files to ingest based on a Gen2 butler query, and that wouldn't be appropriate to move to `RawIngestTask` itself).  That would make the process of bootstrapping a Gen3 repo from a Gen2 one a bit more work, but it would make it a lot more maintainable and especially debuggable.

Reply:
Separating the instrument registration and curated calibrations from conversion has the downside that you now need three commands to convert a single gen2 repo instead of one. Seems annoying. It's also a bit odd that we can't work out the instrument to use from the gen2 mapper but insist on people specifying the gen3 instrument class on the command line.

Reply:
Let's defer the debate on whether to remove the instrument registration and `writeCuratedCalibrations` until later, then - I'm definitely not advocating doing that here.

In terms of fixing this more directly, here's a [link to our Slack conversation](https://lsstc.slack.com/archives/C2JPT1KB7/p1586968385439200) for posterity.  I think the summary is that we agree it's worth doing but I think it can be deferred, probably to [DM-24432](https://jira.lsstcorp.org/browse/DM-24432), which will have to make changes to how this works anyway.